File: src/main/java/br/com/shoppinglist/shopping_list/identity/application/usecase/LoginUserUseCaseImpl.java
package br.com.shoppinglist.shopping_list.identity.application.usecase;

import br.com.shoppinglist.shopping_list.identity.application.dto.AuthTokenResult;
import br.com.shoppinglist.shopping_list.identity.application.dto.LoginUserCommand;
import br.com.shoppinglist.shopping_list.identity.application.port.in.LoginUserUseCase;
import br.com.shoppinglist.shopping_list.identity.application.port.out.TokenIssuer;
import br.com.shoppinglist.shopping_list.identity.application.port.out.UserRepository;
import br.com.shoppinglist.shopping_list.identity.domain.exception.InvalidCredentialsException;
import br.com.shoppinglist.shopping_list.identity.domain.model.User;
import org.springframework.stereotype.Service;

@Service
public class LoginUserUseCaseImpl implements LoginUserUseCase {

	private final UserRepository userRepository;
	private final TokenIssuer tokenIssuer;

	public LoginUserUseCaseImpl(UserRepository userRepository, TokenIssuer tokenIssuer) {
		this.userRepository = userRepository;
		this.tokenIssuer = tokenIssuer;
	}

	@Override
	public AuthTokenResult execute(LoginUserCommand command) {
		String email = normalizeEmail(command.email());
		if (email.isBlank() || command.password() == null || command.password().isBlank()) {
			throw new InvalidCredentialsException();
		}

		User user = userRepository.authenticate(email, command.password())
			.orElseThrow(InvalidCredentialsException::new);

		return tokenIssuer.issue(user.id());
	}

	private String normalizeEmail(String email) {
		if (email == null) {
			return "";
		}
		return email.trim().toLowerCase();
	}
}

File: src/identity/application/usecase/login-user.use-case.impl.ts
import { Injectable } from '@nestjs/common';
import { InvalidCredentialsException } from '../../domain/exception/identity.exceptions';
import { AuthTokenResult } from '../dto/auth-token.result';
import { LoginUserCommand } from '../dto/login-user.command';
import { LoginUserUseCase } from '../port/in/login-user.use-case';
import { PasswordHasher } from '../port/out/password-hasher';
import { TokenIssuer } from '../port/out/token-issuer';
import { UserRepository } from '../port/out/user.repository';
import { RegistrationInputValidator } from './registration-input.validator';

@Injectable()
export class LoginUserUseCaseImpl extends LoginUserUseCase {
  constructor(
    private readonly users: UserRepository,
    private readonly passwordHasher: PasswordHasher,
    private readonly tokenIssuer: TokenIssuer,
  ) {
    super();
  }

  async execute(command: LoginUserCommand): Promise<AuthTokenResult> {
    const email = RegistrationInputValidator.normalizeEmail(command.email);
    if (!email || !command.password?.trim()) {
      throw new InvalidCredentialsException();
    }

    const found = await this.users.findByEmailWithPasswordHash(email);
    if (!found) {
      throw new InvalidCredentialsException();
    }

    const valid = await this.passwordHasher.matches(
      command.password,
      found.passwordHash,
    );
    if (!valid) {
      throw new InvalidCredentialsException();
    }

    return this.tokenIssuer.issue(found.user.id);
  }
}
